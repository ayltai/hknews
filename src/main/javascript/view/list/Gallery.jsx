import PropTypes from 'prop-types';
import React from 'react';
import { SingleFieldList, } from 'react-admin';
import Lightbox from 'react-image-lightbox';
import 'react-image-lightbox/style.css';

import { StringUtils, } from '../../util/StringUtils';

const getEntries = props => props.ids && props.data ? props.ids.map(id => props.data[id]) : [];

const getImages = props => Object.entries(getEntries(props)).map(record => ({
    label  : StringUtils.decode(record[1][props.caption]),
    source : record[1][props.source],
}));

export const Gallery = props => {
    const [ index, setIndex, ]      = React.useState(0);
    const [ open,  setOpen,  ]      = React.useState(false);
    const { source, ...otherProps } = props;
    const entries                   = getEntries(props);
    const images                    = getImages(props);

    const children = React.Children.map(props.children, child => React.cloneElement(child, {
        onClick : event => {
            setIndex(Math.max(0, entries ? entries.findIndex(entry => entry[source] === event.target.src) : 0));
            setOpen(true);
        },
    }));

    return (
        <React.Fragment>
            <SingleFieldList
                linkType={false}
                {...otherProps}>
                {children[0]}
            </SingleFieldList>
            {open && (
                <Lightbox
                    mainSrc={images[index].source}
                    nextSrc={index + 1 < images.length ? images[(index + 1) % images.length].source : undefined}
                    prevSrc={index - 1 >= 0 ? images[(index + images.length - 1) % images.length].source : undefined}
                    imageCaption={images[index].label}
                    onCloseRequest={() => setOpen(false)}
                    onMoveNextRequest={() => setIndex((index + 1) % images.length)}
                    onMovePrevRequest={() => setIndex((index + images.length - 1) % images.length)} />
            )}
        </React.Fragment>
    );
};

Gallery.propTypes = {
    ids      : PropTypes.arrayOf(PropTypes.number),
    data     : PropTypes.object,
    caption  : PropTypes.string,
    source   : PropTypes.string,
    children : PropTypes.object,
};
