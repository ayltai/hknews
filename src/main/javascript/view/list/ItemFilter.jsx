import React from 'react';
import { Filter, ReferenceInput, SelectInput, } from 'react-admin';

export const ItemFilter = props => (
    <Filter {...props}>
        <ReferenceInput
            label='labels.publisher'
            source='source'
            reference='sources'
            allowEmpty
            alwaysOn>
            <SelectInput
                optionText='name'
                variant='standard' />
        </ReferenceInput>
    </Filter>
);
