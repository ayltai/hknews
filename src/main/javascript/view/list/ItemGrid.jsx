import { Card, CardActions, CardActionArea, CardContent, CardHeader, CardMedia, Grid, makeStyles, Tooltip, } from '@material-ui/core';
import PropTypes from 'prop-types';
import React from 'react';
import { RichTextField, ShowButton, TextField, } from 'react-admin';
import TimeAgo from 'react-timeago';

import { StringUtils, } from '../../util/StringUtils';
import { Constants, } from '../../Constants';

export const ItemGrid = ({ basePath, ids, data, }) => {
    const classes = makeStyles(theme => ({
        container   : {
            padding  : theme.spacing(2),
            flexGrow : 1,
        },
        card        : {
            display : 'flex',
        },
        details     : {
            display       : 'flex',
            flexDirection : 'column',
            flexGrow      : 1,
        },
        content     : {
            flex : '1 0 auto',
        },
        title       : {
            display         : '-webkit-box',
            overflow        : 'hidden',
            fontSize        : '120%',
            fontWeight      : 'bold',
            textOverflow    : 'ellipsis',
            whiteSpace      : 'break-spaces',
            WebkitBoxOrient : 'vertical',
            WebkitLineClamp : 1,
        },
        description : {
            minHeight       : theme.spacing(12),
            display         : '-webkit-box',
            overflow        : 'hidden',
            textOverflow    : 'ellipsis',
            WebkitBoxOrient : 'vertical',
            WebkitLineClamp : 4,
        },
        media       : {
            minWidth   : theme.spacing(20),
            width      : theme.spacing(20),
            minHeight  : theme.spacing(20),
            height     : theme.spacing(20),
            marginLeft : theme.spacing(2),
        },
        actions     : {
            display : 'block',
        },
    }))();

    return (
        <div className={classes.container}>
            <Grid
                container
                spacing={2}>
                {ids.map(id => (
                    <Grid
                        item
                        key={id}
                        lg={12}
                        xl={6}>
                        <Card>
                            <CardActionArea href={`#${basePath}/${id}/show`}>
                                <CardHeader
                                    className={classes.content}
                                    title={
                                        <TextField
                                            record={data[id]}
                                            source='sourceName' />
                                    }
                                    subheader={<TimeAgo date={data[id].publishDate} />}
                                    avatar={
                                        <img
                                            src={`${Constants.API_ENDPOINT}/images/${Constants.SOURCE_IMAGES[data[id].sourceName]}.png`}
                                            width={Constants.AVATAR_SIZE}
                                            height={Constants.AVATAR_SIZE}
                                            title={data[id].sourceName}
                                            alt={data[id].sourceName} />
                                    } />
                                <CardContent className={classes.content}>
                                    <div className={classes.card}>
                                        <div className={classes.details}>
                                            <RichTextField
                                                className={classes.title}
                                                record={data[id]}
                                                source='title' />
                                            <RichTextField
                                                className={classes.description}
                                                record={data[id]}
                                                source='description' />
                                        </div>
                                        {data[id].images && data[id].images.length > 0 && (
                                            <Tooltip title={data[id].images[0].description ? StringUtils.decode(data[id].images[0].description) : ''}>
                                                <CardMedia
                                                    className={classes.media}
                                                    image={data[id].images[0].url} />
                                            </Tooltip>
                                        )}
                                    </div>
                                </CardContent>
                                <CardActions className={classes.actions}>
                                    <ShowButton
                                        basePath={basePath}
                                        label='labels.read_more'
                                        record={data[id]} />
                                </CardActions>
                            </CardActionArea>
                        </Card>
                    </Grid>
                ))}
            </Grid>
        </div>
    );
};

ItemGrid.propTypes = {
    basePath : PropTypes.string,
    ids      : PropTypes.arrayOf(PropTypes.number),
    data     : PropTypes.object,
};
