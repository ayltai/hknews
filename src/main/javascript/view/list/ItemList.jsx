import React from 'react';
import { List, Pagination, } from 'react-admin';

import { Constants, } from '../../Constants';
import { ScrollToTopOnMount, } from '../ScrollToTopOnMount';
import { ItemFilter, } from './ItemFilter';
import { ItemGrid, } from './ItemGrid';

export const ItemList = props => (
    <React.Fragment>
        <ScrollToTopOnMount />
        <List
            bulkActionButtons={false}
            exporter={false}
            filters={<ItemFilter />}
            perPage={Constants.NEWS_PER_PAGE}
            pagination={<Pagination rowsPerPageOptions={Constants.NEWS_PER_PAGE_OPTIONS} />}
            {...props}>
            <ItemGrid />
        </List>
    </React.Fragment>
);
