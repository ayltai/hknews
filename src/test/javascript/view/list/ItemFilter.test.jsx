import { mount, } from 'enzyme';
import React from 'react';
import { TestContext, } from 'ra-test';
import renderer from 'react-test-renderer';

import { ItemFilter, } from '../../../../main/javascript/view/list/ItemFilter';

const component = (
    <TestContext>
        <ItemFilter
            resource='source'
            displayFilters={{}}
            hideFilter={() => {
                return;
            }} />
    </TestContext>
);

describe('<ItemFilter />', () => {
    it('renders without errors', () => {
        expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('mounts without errors', () => {
        mount(component);
    });
});
