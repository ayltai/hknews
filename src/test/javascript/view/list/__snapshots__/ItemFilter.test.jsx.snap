// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemFilter /> renders without errors 1`] = `
<form
  __versions={
    Object {
      "final-form": "4.20.2",
      "react-final-form": "6.5.2",
    }
  }
  className="RaFilter-form-2 RaFilterForm-form-3"
  displayFilters={Object {}}
  onSubmit={[Function]}
>
  <div
    className="filter-field RaFilterFormInput-body-5"
    data-source="source"
  >
    <div
      className="MuiFormControl-root MuiFormControl-marginDense"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root RaLabeled-label-9 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense"
        data-shrink={true}
        htmlFor="source"
      >
        <span>
          labels.publisher
        </span>
      </label>
      <div
        className="RaLabeled-value-10"
      />
    </div>
    <div
      className="RaFilterFormInput-spacer-6"
    >
      Â 
    </div>
  </div>
  <div
    className="RaFilterForm-clearFix-4"
  />
</form>
`;
